File: swift/88_MergeSortedArray.swift
class Solution {
    func merge(_ nums1: inout [Int], _ m: Int, _ nums2: [Int], _ n: Int) {
        var i = m - 1, j = n - 1
        for (idx, n) in nums1.enumerated().reversed() {
            if i >= 0 && j >= 0 {
                if nums1[i] > nums2[j] {
                    nums1[idx] = nums1[i]
                    i -= 1
                } else {
                    nums1[idx] = nums2[j]
                    j -= 1
                }
            } else {
                if i >= 0 {
                    nums1[idx] = nums1[i]
                    i -= 1
                } else {
                    nums1[idx] = nums2[j]
                    j -= 1
                }
            }
        }
    }
}

File: cpp/88_MergeSortedArray.cpp
class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int idx = m + n - 1;
        m--, n--;
        
        while (idx >= 0) {
            if (m >= 0 && n >= 0) {
                nums1[idx] = nums1[m] > nums2[n] ? nums1[m--] : nums2[n--];
            } else {
                nums1[idx] = n >= 0 ? nums2[n--] : nums1[m--];
            }
            idx--;
        }
    }
};

class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int idx = n+m-1;
        m--, n--;
        while (idx >= 0) {
            if (n < 0 && m < 0) {
                return;
            } else if (n >= 0 && m >= 0) {
                nums1[idx--] = nums1[m] > nums2[n] ? nums1[m--] : nums2[n--];
            } else {
                nums1[idx--] = m >= 0 ? nums1[m--] : nums2[n--];
            }
        }
    }
};

class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int idx = 0;
        for (int i = 0; i < n+m; ++i) {
            if (idx < n) {
                if (nums1[i] > nums2[idx]) {
                    for (int j = n+m-2; j >= i; --j) nums1[j+1] = nums1[j];
                    nums1[i] = nums2[idx++];
                } else if (i >= m+idx) {
                    while (idx < n) nums1[i++] = nums2[idx++];
                }
            } else {
                return;
            }
        }
    }
};
